scanner/matcher/updater/updater.go

```diff
@@ -0,0 +1,364 @@
+
+	return u, nil
+}
+
+func fillOpts(opts *Opts) error {
+	if opts == nil {
+		panic("programmer error")
+	}
+	if err := validate(*opts); err != nil {
+		return err
+	}
+
+	if opts.Client == nil {
+		opts.Client = http.DefaultClient
+	}
+	if opts.Root == "" {
+		opts.Root = os.TempDir()
+	}
+	if opts.UpdateInterval < time.Minute {
+		opts.UpdateInterval = defaultUpdateInterval
+	}
+	if opts.UpdateRetention <= 1 {
+		opts.UpdateRetention = defaultUpdateRetention
+	}
+	if opts.FullGCInterval < 3*time.Hour {
+		opts.FullGCInterval = defaultFullGCInterval
+	}
+
+	return nil
+}
+
+func validate(opts Opts) error {
+	if opts.Store == nil || opts.Locker == nil || opts.Pool == nil || opts.MetadataStore == nil {
+		return errors.New("must provide a Store, a Locker, a Pool, and a MetadataStore")
+	}
+	if _, err := url.Parse(opts.URL); err != nil {
+		return fmt.Errorf("invalid URL: %q", opts.URL)
+	}
+	return nil
+}
+
+// tryRemoveExiting attempts to remove any leftover update files present in the updater's
+// root directory.
+//
+// It is assumed the previous updater used the same root directory.
+// Any errors when attempting to remove pre-existing files are simply logged.
+func (u *Updater) tryRemoveExiting() {
+	ctx := zlog.ContextWithValues(u.ctx, "component", "matcher/updater/Updater.tryRemoveExiting")
+
+	files, err := fs.Glob(os.DirFS(u.root), updateFilePattern)
+	if err != nil {
+		zlog.Warn(ctx).Err(err).Msg("searching for previous update files to remove")
+	}
+	for _, file := range files {
+		if err := os.Remove(filepath.Join(u.root, file)); err != nil {
+			zlog.Warn(ctx).Err(err).Msgf("removing previous update file %s", file)
+		}
+	}
+}
+
+// Stop cancels the updater's context and prevents future update cycles.
+// It does **not** cleanup any other resources. It is the user's responsibility to do so.
+func (u *Updater) Stop() error {
+	u.cancel()
+	return nil
+}
+
+// Start periodically updates the vulnerability data.
+// Each period is adjusted by some amount of jitter.
+func (u *Updater) Start() error {
+	ctx := zlog.ContextWithValues(u.ctx, "component", "matcher/updater/Updater.Start")
+
+	zlog.Info(ctx).Msg("starting initial update")
+	if err := u.update(ctx); err != nil {
+		zlog.Error(ctx).Err(err).Msg("errors encountered during updater run")
+	}
+	zlog.Info(ctx).Msg("completed initial update")
+
+	if !u.skipGC {
+		go u.runGCFullPeriodic()
+	}
+
+	timer := time.NewTimer(u.updateInterval + jitter())
+	defer timer.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-timer.C:
+			zlog.Info(ctx).Msg("starting update")
+			if err := u.update(ctx); err != nil {
+				zlog.Error(ctx).Err(err).Msg("errors encountered during updater run")
+			}
+			zlog.Info(ctx).Msg("completed update")
+
+			timer.Reset(u.updateInterval + jitter())
+		}
+	}
+}
+
+// update runs the full vulnerability update process.
+func (u *Updater) update(ctx context.Context) error {
+	if err := u.runUpdate(ctx); err != nil {
```

Comment:
consider returning a bool to indicate if the update actually happened. perhaps don't bother running GC if not update occurred